core/src/main/scala/kafka/server/AbstractFetcherManager.scala

```diff
@@ -66,6 +67,15 @@ abstract class AbstractFetcherManager[T <: AbstractFetcherThread](val name: Stri
   Map("clientId" -> clientId)
   )
 
+  newGauge(
+    "FailedPartitionsCount", {
+      new Gauge[Long] {
+        def value = failedPartitions.size
```

Comment:
We might need to be a little careful here. We are using the lock elsewhere to protect access to `failedPartitions`, but this is unprotected. I think the only harm from this at the moment is that we may see a stale size for this metric. If we want to be on the safe side, we could maintain the size as a separate field (e.g. using an `AtomicInteger`).

Reply:
The synchronized functions in the FailedPartition class would take care of this now.